### PR TITLE
Allow to specify assembly scanner to scan nested folder

### PR DESCRIPTION
**Pull request by SeanFeldman**
### Breaking changes

Changing default AssemblyScanner behavior **not** to scan nested folders.
Nested folders will be scanned **only** when requested so through code API.
~~To support Azure and NSB hosts, a command line argument will be added (`-scanNested`).~~

Things to do:
- [x] Change to `BusConfiguration` to support enabling nested directories scanning
- [ ] Update documention

*Update*: summary of why code API is the only scan we care about is [here](https://github.com/Particular/NServiceBus/pull/2711#issuecomment-93460849)

### CODE REVIEW

**Comment by SeanFeldman** [comment]
@Particular/nservicebus please review

**Comment by johnsimons** [comment]
@SeanFeldman I think it would cleaner to new up a new appdomain and auto gen a config file on the fly, thoughts ?

**Comment by danielmarbach** [comment]
@SeanFeldman see the discussion happening here https://github.com/Particular/NServiceBus.RavenDB/pull/43#discussion_r28228230

We should build a code first API.

**Comment by danielmarbach** [comment]
So you plan to introduce a behavior breaking change because by default now it doesn't scan nested directories. I would argue that we should leave the default to scan nested directories and a code centric config API to override that behavior. I added the breaking change label for now

**Comment by SimonCropp** [comment]
> I would argue that we should leave the default to scan nested directories and a code centric config API to override that behavior.

+1

**Comment by SimonCropp** [comment]
also no value this being a config setting. just make it code only

**Comment by danielmarbach** [comment]
@SimonCropp you could have +1 on https://github.com/Particular/NServiceBus/pull/2711#issuecomment-92669386 ;)

**Comment by johnsimons** [comment]
The breaking change has been planned for a while, scanning nested folders does not make sense, hence the reason to change it to not be the default and instead make it an opt in.

Regarding the appsetting, @SeanFeldman and I discussed it, the issue is that our host currently also used the AssemblyScanner class to do the first scanning, and that scanning is done very early on, it is used to find `IConfigureThisEndpoint`, so the only way for the users to control that scanning would be if we either make it a host passed in argument eg -scanDeep=true or the way we did it.
We thought that adding a code config option to assembly scan could confuse users since that setting is read too late because of the initial scanning mentioned above. 

Anyway, if you guys can suggest a better way that also control the initial host scanning, go for it.

**Comment by SimonCropp** [comment]
prefer putting `-scanDeep=true ` in the host

**Comment by SimonCropp** [comment]
> The breaking change has been planned for a while, scanning nested folders does not make sense, hence the reason to change it to not be the default and instead make it an opt in.

Fair enough. will need to make the very clear in the upgrade guide

**Comment by johnsimons** [comment]
We can do that but then it feels confusing IMO because how would u
configure it for selfhosting?

**Comment by SimonCropp** [comment]
>how would u configure it for selfhosting

using the code API

**Comment by andreasohlund** [comment]
+1 for code first api only like @danielmarbach mentions

@SeanFeldman can you update the description to reflect this?

**Comment by andreasohlund** [comment]
@SeanFeldman I've summarized the code first discussion into a "Design principle"  https://github.com/Particular/Vision/issues/8

**Comment by SeanFeldman** [comment]
Switching to code only and providing an argument `-scanDeep=true` will require to maintain this logic for 2 hosts - regular NSB host and Azure NSB host.

**Comment by andreasohlund** [comment]
>Switching to code only and providing an argument -scanDeep=true will require to maintain this logic for 2 hosts - regular NSB host and Azure NSB host.

I'd say this is temporary since we're deprecating the host in favor of the bootstrap approach anyway?

**Comment by SeanFeldman** [comment]
>  deprecating the host in favour of the bootstrap approach anyway?
Thank you about remind that @andreasohlund :)

I'd also suggest to come up with a more intention revealing name for the command line argument. Perhaps `-scanNested` instead of `-scanDeep`?

**Comment by SeanFeldman** [comment]
@johnsimons since we're going away from config file based setting, the test will not require it. But I'd like to see [how](https://github.com/Particular/NServiceBus/pull/2711#issuecomment-92633767) that test could be written.

**Comment by SeanFeldman** [comment]
There are two places with NSB host that instantiate `AssemblyScanner`
1. [`HostProgram`](https://github.com/Particular/NServiceBus.Host/blob/develop/src/NServiceBus.Hosting.Windows/HostProgram.cs#L32)
2. [`GenericHost`](https://github.com/Particular/NServiceBus.Host/blob/develop/src/NServiceBus.Hosting.Windows/GenericHost.cs#L32)
Any particular reason to do so @johnsimons ?

**Comment by johnsimons** [comment]
@SeanFeldman the second scan in `GenerisHost` is needed if we did not do a scan at `HostProgram` because for example the user specified the exact location for `IConfigureThisEndpoint`

**Comment by SeanFeldman** [comment]
Reviewed assembly scanning with @andreasohlund and here are updates:

- With the new default (*not* scanning nested directories by default), we won't need command line arguments. 
  - For [`HostProgram`](https://github.com/Particular/NServiceBus.Host/blob/develop/src/NServiceBus.Hosting.Windows/HostProgram.cs#L32) assembly scanning is only performed to locate the type that is configuring executing endpoint.
  - For [`GenericHost`](https://github.com/Particular/NServiceBus.Host/blob/develop/src/NServiceBus.Hosting.Windows/GenericHost.cs#L32) assembly scanning is used to [find profiles](https://github.com/Particular/NServiceBus.Host/blob/develop/src/NServiceBus.Hosting.Windows/GenericHost.cs#L45) and later those assemblies are [passed to user code](https://github.com/Particular/NServiceBus.Host/blob/develop/src/NServiceBus.Hosting.Windows/GenericHost.cs#L118) through `BusConfiguration` instance where user can override default and set nested directories to be scanned. If a profile is found in a nested folder, we could document the fact that it shouldn't, or instruct how to create a profile that instructs to search in nested folders (@andreasohlund - please confirm).
- **Host Issue**: Currently `GenericHost` scans assemblies and passes into `BusConfiguration` using [AssembliesToScan() method](https://github.com/Particular/NServiceBus.Host/blob/develop/src/NServiceBus.Hosting.Windows/GenericHost.cs#L110) which is deprecated and throwing exception. This call shouldn't take place at all as we don't really care about assemblies scanned by generic host and can override those with user code.

**Comment by andreasohlund** [comment]
>For HostProgram assembly scanning is only performed to locate the type that is configuring executing endpoint.

Correct, and if that type is in a nested folder we can guide the user to specify the type explicitly

>For GenericHost assembly scanning is used to find profiles and later those assemblies are passed to user code through BusConfiguration instance where user can override default and set nested directories to be scanned. If a profile is found in a nested folder, we could document the fact that it shouldn't, or instruct how to create a profile that instructs to search in nested folders (@andreasohlund - please confirm).

Correct, If a user for some reason have a profile in a nested folder we can document that he needs to move it.

**Comment by johnsimons** [comment]
The only down point is that before we were able to perform only a single scan, and with the current proposal there will always be more then one scan when using our host, the first by the host and the second by BusConfiguration, but this is a different issue that we may be able to solve later on.

**Comment by SeanFeldman** [comment]
Wanted to add an acceptance test for this, but could due to the fact that `AssemblyScanner` is used by ATT and going through `BusConfiguration` is not working. Suggestions are welcomed @Particular/nservicebus

**Comment by SeanFeldman** [comment]
Tests are updated to go use `BusConfiguration` and not `AsseblyScanner` directly.

**Comment by SeanFeldman** [comment]
Should this be merged to `develop` @Particular/nservicebus ?

**Comment by andreasohlund** [comment]
Needs a rebase

**Comment by johnsimons** [comment]
:+1: to merge

File: src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs [hunk 1]
```diff
@@ -70,6 +72,16 @@ internal AssemblyScanner(Assembly assemblyToScan)
             ThrowExceptions = true;
         }
 
+        void DetermineIfNestedDirectoriesShouldBeExcluded()
+        {
+            bool scanNestedDirectories;
+            var appSetting = ConfigurationManager.AppSettings["NServiceBus/AssemblyScanning/ScanNestedDirectories"];
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs]: disagree with having an app setting here. why is this required?

**Comment by SimonCropp** [comment]
if the host needs an app setting then put that logic in the host. the host can then use the code api

**Comment by SimonCropp** [comment]
is there any associated doco?

File: src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs]: +1, @SeanFeldman didn't we come to the conclusion that we didn't need it?

Comment (SeanFeldman) [inline src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs]: @andreasohlund I'll have to look into this. It was changed to rely on code only config...

**Comment by SeanFeldman** [comment]
@SimonCropp removed the appConfig reference that was stale.
@andreasohlund rebased to the recent `develop`.
Will create a PR for doco and then we can merge this one.

**Comment by SeanFeldman** [comment]
How should we approach documentation? 
Would we create a doco branch for V6 were all documentation changes for v6 would go? 

@Particular/documentation what do you think?

**Comment by andreasohlund** [comment]
Talked about this with @SimonCropp yesterday. Here is our proposal

* We should start to create vN+1 docs right away(as soon as we have decided to bump dev) to not incure "doco" debt. The first step is to update samples snippets after all obsoletes have been removed
* Simon will add "exclude" support to the site so we can decide when snippets for vN+1 starts to show up
* Upgrade guides should be worked on right away since there is no harm in them being public right away
* We'll either check in the vN+1 assembly or add it to myget as needed to get it compiling

What do you think?

If yes we need to get going to create the snippets for v6 and create the upgrade guide as we do it (since upgrading the snippets is a very good way to do it)

**Comment by SeanFeldman** [comment]
+1 @andreasohlund for going this route. Having doco sitting in a PR would generate too much work of merging and conflict resolving later.

@simoncropp can I suggest to have a query string that would allow rendering v6 forv"internal preview" needs?

**Comment by SimonCropp** [comment]
Why do u need to preview it?

**Comment by SeanFeldman** [comment]
The real thing when rendered is not the same as presented in MarkdownPad.

**Comment by SeanFeldman** [comment]
Merging this to be able to update doco with snippets. Can't reference PR nuget as it then breaks other another v6 snippets that were referencing another PR.

**Comment by SeanFeldman** [comment]
Closes #2708
